### ISSUE TITLE
Bug: Order of conditional format

### ISSUE THREAD

**Issue by A-Hegazi** [jmcnamara/rust_xlsxwriter#113]
### Question

Hello,
I have question about order of conditional format
I'm trying to make a new excel sheet wth 20 conditional format
and they are overlapping in some cells but when i open the sheet the order is reversed i tried to reverse the range in my code `for i in (0..19).rev() {` but it did not work also i tried to use hard code not a loop and played with the order in the code and it aalways give me the same outcome am i doing some thing wrong?
```
    for i in 0..19 {
        let form = Format::new()
            .set_background_color(Color::RGB(color_array[i].0))
            .set_font_color(Color::RGB(color_array[i].1))
            .set_border_color(Color::RGB(BORDER_COLOR))
            .set_border_top(FormatBorder::Thin)
            .set_border_bottom(FormatBorder::Thin)
            .set_bold()
            .set_align(FormatAlign::Center);

        let rule = format!("$W4>={}", i + 1);
        let conditional = ConditionalFormatFormula::new()
            .set_rule(rule.as_str())
            .set_format(form);
        worksheet.add_conditional_format(3, i as u16, MAX_ROWS, 22, &conditional)?;
    }
```
![image](https://github.com/user-attachments/assets/709bc978-8688-4cf0-ae72-3982c1ef809d)

**Comment by jmcnamara** [comment]
The order that the conditional formats are applied should be the same as the order they are added in.

Could you please add a small complete working program and I'll have a look at it. It probably doesn't need 20 conditional formats to reproduce it. Try with 5 or 6.

**Comment by A-Hegazi** [comment]
*This is the desired out come
![image](https://github.com/user-attachments/assets/5b29ff59-d00b-4ce1-b102-c628e5ee66ce)
*and this is the complete project

```
use rust_xlsxwriter::*;
fn main() -> Result<(), XlsxError> {
    const MAX_ROWS: u32 = 5000;
    //Colors
    let color_array: [(u32, u32); 20] = [
        (0x0000ff, 0xffff00),
        (0x80ff80, 0x000000),
        (0xffff00, 0x3a00ff),
        (0x0000ff, 0xffffff),
        (0xff0000, 0xffffff),
        (0x80ffff, 0x000000),
        (0xff80ff, 0x000000),
        (0xffff80, 0x000000),
        (0x000000, 0xffffff),
        (0xc0c0c0, 0xffffff),
        (0x008000, 0xffffff),
        (0x0000a0, 0xffffff),
        (0x804000, 0xffffff),
        (0x800080, 0xffffff),
        (0xff8040, 0xffffff),
        (0x8080c0, 0xffffff),
        (0x808040, 0xffffff),
        (0x808080, 0xffffff),
        (0x408080, 0xffffff),
        (0x8080c0, 0xffffff),
    ];

    const BORDER_COLOR: u32 = 0xc0c0c0;
    // Create a new Excel file object.
    let mut workbook = Workbook::new();

    // Add a worksheet to the workbook.
    let worksheet = workbook.add_worksheet();

    //Set worksheet name
    worksheet.set_name("WBS")?;

    //Add heading data
    worksheet.write(2, 20, "WBS Code")?;
    worksheet.write(2, 21, "WBS Name")?;
    worksheet.write(2, 22, "WBS Level")?;

    //Add first level data
    worksheet.write(3, 20, "Project Code")?;
    worksheet.write(3, 21, "Project Name")?;
    worksheet.write(3, 22, "1")?;

    //first level format *Complete Later*
    let first_level_format = Format::new()
        .set_background_color(Color::RGB(color_array[0].0))
        .set_font_color(Color::RGB(color_array[0].1))
        .set_font_size(12)
        .set_border_color(Color::RGB(BORDER_COLOR))
        .set_border_top(FormatBorder::Thin)
        .set_border_bottom(FormatBorder::Thin)
        .set_bold();
    worksheet.set_range_format(3, 0, 3, 22, &first_level_format)?;

    //special first level format *Complete Later*
    let s_first_level_format = Format::new()
        .set_background_color(Color::RGB(color_array[0].0))
        .set_font_color(Color::RGB(color_array[0].1))
        .set_font_size(12)
        .set_border_color(Color::RGB(BORDER_COLOR))
        .set_border_top(FormatBorder::Thin)
        .set_border_left(FormatBorder::Thin)
        .set_bold();
    worksheet.set_cell_format(3, 0, &s_first_level_format)?;

    //Conditional format 1
    for i in 0..19 {
        // println!("{i}");
        let form = Format::new()
            .set_background_color(Color::RGB(color_array[i].0))
            .set_font_color(Color::RGB(color_array[i].1))
            .set_border_color(Color::RGB(BORDER_COLOR))
            .set_border_top(FormatBorder::Thin)
            .set_border_bottom(FormatBorder::Thin)
            .set_bold()
            .set_align(FormatAlign::Center);

        let rule = format!("$W4>={}", i + 1);
        let conditional = ConditionalFormatFormula::new()
            .set_rule(rule.as_str())
            .set_format(form);
        worksheet.add_conditional_format(3, i as u16, MAX_ROWS, 22, &conditional)?;
    }

    //set  columns width
    worksheet.set_column_range_width(0, 19, 0.5)?;
    worksheet.set_column_range_width(20, 22, 13)?;
    worksheet.set_column_width(21, 50)?;

    // Save the file to disk.
    workbook.save("test.xlsx")?;

    Ok(())
}

```

**Comment by A-Hegazi** [comment]
after digging into the problem i tried to see the root cause to see what factor affect it

i think it relate to the `add_conditional_format() ` i tried to make rule for range that nested inside other range that also has conditional format see the below figure
![image](https://github.com/user-attachments/assets/c9edadac-54a3-42c3-abf5-8334dbfb6b68)
when i tried this code
```
    let format1 = Format::new().set_background_color(Color::Red);
    let rule1 = format!("$W4>={}", 1);
    let conditional1 = ConditionalFormatFormula::new()
        .set_rule(rule1.as_str())
        .set_format(format1);
    worksheet.add_conditional_format(0, 0, 5, 5, &conditional1)?;

    let format2 = Format::new().set_background_color(Color::Blue);
    let rule2 = format!("$W4>={}", 2);
    let conditional2 = ConditionalFormatFormula::new()
        .set_rule(rule2.as_str())
        .set_format(format2);
    worksheet.add_conditional_format(0, 1, 5, 5, &conditional2)?;
```

here you will see the Red rule come before the Blue but the Blue is nested inside it and the outcome is:
![image](https://github.com/user-attachments/assets/2c31064c-0e46-4960-bc6c-0fc9a63b7e5f)
if we changed the range without change the order of the code just making the Red inside the Blue
```
    let format1 = Format::new().set_background_color(Color::Red);
    let rule1 = format!("$W4>={}", 1);
    let conditional1 = ConditionalFormatFormula::new()
        .set_rule(rule1.as_str())
        .set_format(format1);
    worksheet.add_conditional_format(0, 1, 5, 5, &conditional1)?;

    let format2 = Format::new().set_background_color(Color::Blue);
    let rule2 = format!("$W4>={}", 2);
    let conditional2 = ConditionalFormatFormula::new()
        .set_rule(rule2.as_str())
        .set_format(format2);
    worksheet.add_conditional_format(0, 0, 5, 5, &conditional2)?;
```

the out come will be:
![image](https://github.com/user-attachments/assets/6011b860-0fae-4a0a-b72a-3510fdb8d12e)
it does not follow the order of the code block it follow who is inside who

but in fact if the two rules have the same range it follow the order of the code
```
    let format1 = Format::new().set_background_color(Color::Red);
    let rule1 = format!("$W4>={}", 1);
    let conditional1 = ConditionalFormatFormula::new()
        .set_rule(rule1.as_str())
        .set_format(format1);
    worksheet.add_conditional_format(0, 0, 5, 5, &conditional1)?;

    let format2 = Format::new().set_background_color(Color::Blue);
    let rule2 = format!("$W4>={}", 2);
    let conditional2 = ConditionalFormatFormula::new()
        .set_rule(rule2.as_str())
        .set_format(format2);
    worksheet.add_conditional_format(0, 0, 5, 5, &conditional2)?;
```
![image](https://github.com/user-attachments/assets/94b01388-3458-40a4-9efa-46eeeb408736)
i don't know why it behave like this i tried to help as i can but i can't figure out this behavior

**Comment by jmcnamara** [comment]
From a first look this seems to be standard Excel behavior. Do you get the same or different behavior if you create the rules in Excel?

**Comment by A-Hegazi** [comment]
Actully in excel , the order of the rules doesn't depend on the range size or nesting. It's determined solely by the order in which you create them. 

The last rule you create will be on the top and have the highest priority.

**Comment by A-Hegazi** [comment]
Actully in excel , the order of the rules doesn't depend on the range size or nesting. It's determined solely by the order in which you create them. 

The last rule you create will be on the top and have the highest priority.

**Comment by jmcnamara** [comment]
> Actully in excel , the order of the rules doesn't depend on the range size or nesting. It's determined solely by the order in which you create them.

In that case could you attach a file with manually applied rules that reflects your desired output.

**Comment by A-Hegazi** [comment]
> In that case could you attach a file with manually applied rules that reflects your desired output and mimics the file you are trying to generate in code.

actually i want to say thank you for your strong support second sure i will attach a file 
[example.xlsx](https://github.com/user-attachments/files/17576082/example.xlsx)

what i want to achieve is some thing like this 
![image](https://github.com/user-attachments/assets/ceb558ee-adc8-4266-a801-8709061648ea)

**Comment by jmcnamara** [comment]
> what i want to achieve is some thing like this

Okay. I think I see the issue now and this is probably a bug. 

The conditional formats are being sorted by the range whereas you expect (reasonably I think) that the conditional formats appear in the order or creation (or the reverse order of creation). I'll look at putting a fix in for that.

**Comment by A-Hegazi** [comment]
Thank you for your consideration

**Comment by jmcnamara** [comment]
I've pushed a fix for this to main.

You will need to use `for i in (0..19).rev() {}` since the ordering corresponds to the order the conditional formats are added in.

Here is the output from your sample code with the WBS numbers filled in manually:

![screenshot 1](https://github.com/user-attachments/assets/c6107298-b878-4cde-b25a-ec83c439dfe4)

Try it out and let me know how you get on.

**Comment by A-Hegazi** [comment]
i tried now to reverse the loop but i got the same old behavior
![image](https://github.com/user-attachments/assets/77fdc1a8-0bef-4ea6-a745-35a120694ceb)
![image](https://github.com/user-attachments/assets/29242cc4-6740-4ad1-9911-805e6caf20ce)
do i need to change some thing in the crate it self first?

**Comment by jmcnamara** [comment]
> do i need to change some thing in the crate it self first?

Yes. Until the crate is published you will need to test with the GitHub version by adding it to your `Cargo.toml` like this

```
cargo add --git https://github.com/jmcnamara/rust_xlsxwriter.git rust_xlsxwriter
```

**Comment by A-Hegazi** [comment]
It work perfectly !, thank you :)

**Comment by jmcnamara** [comment]
Great. I'll release it in a crate (probably 0.79.1) later today.

**Comment by jmcnamara** [comment]
This fix is now upstream in v0.79.1. Thanks for the report.
